code/modules/projectiles/guns/energy/railgun.dm

```diff
@@ -156,3 +156,109 @@
+	icon = 'icons/obj/guns/energy/gauss.dmi'
+	icon_state = "gauss"
+	item_state = "gauss"
+	fire_sound = 'sound/weapons/guns/fire/gaussrifle.ogg'
+	w_class = ITEM_SIZE_HUGE
+	matter = list(MATERIAL_PLASTEEL = 40, MATERIAL_STEEL = 15, MATERIAL_SILVER = 10, MATERIAL_GOLD = 6)
+	charge_cost = 1000
+	fire_delay = 30
+	recoil_buildup = 30
+	one_hand_penalty = 80 //guh
+	zoom_factor = 1.8
+	extra_damage_mult_scoped = 0.2
+	damage_multiplier = 1.6
+	penetration_multiplier = 2.0
+	twohanded = TRUE
+	slowdown_hold = 1.5
+	brace_penalty = 30
+	init_firemodes = list(
+		list(mode_name="powered-rod", mode_desc="fires a metal rod at incredible speeds", projectile_type=/obj/item/projectile/plasma/gauss, icon="kill"),
+		list(mode_name="fragmented scrap", mode_desc="fires a brittle, sharp piece of scrap-metal", projectile_type=/obj/item/projectile/bullet/grenade/frag, charge_cost=30000, icon="grenade"),
+	)
+	consume_cell = FALSE
+	price_tag = 6000
+
+	var/max_stored_matter = 4
+	var/stored_matter = 0
+	var/matter_type = MATERIAL_RSCRAP
+
+	var/projectile_cost = 1
+	var/overheat_damage = 25
+
+/obj/item/gun/energy/laser/railgun/gauss/Initialize()
+    ..()
+    AddComponent(/datum/component/heat, COMSIG_CLICK_CTRL, TRUE,  50,  60,  20, 0.01, 2)
+    RegisterSignal(src, COMSIG_HEAT_VENT, .proc/ventEvent) //this sould just be a fluff message, proc can be anything
+    RegisterSignal(src, COMSIG_HEAT_OVERHEAT, .proc/handleoverheat) //this can damge the user/melt the gun/whatever. this will never proc as the gun cannot fire above the special heat threshold and the special heat threshold should be smaller than the overheat threshold
+
+/obj/item/gun/energy/laser/railgun/gauss/attackby(obj/item/I, mob/user)
+
+	if(istype(I,/obj/item/gun_upgrade/mechanism/battery_shunt))
+		to_chat(user, SPAN_WARNING("[src] cannot be fitted with a battery shunt!"))		//No shunts on this gun. It's powerful enough.
+		return FALSE
+	else if(!istype(I,/obj/item/stack/sheet))
+		..()
+	var/obj/item/stack/sheet/M = I
+	if(istype(M) && M.name == matter_type)
+		var/amount = min(M.get_amount(), round(max_stored_matter - stored_matter))
+		if(M.use(amount))
+			stored_matter += amount
+		to_chat(user, "<span class='notice>You load [amount] [matter_type] into \the [src].</span>")
+	else
+		return ..()
+
+/obj/item/gun/energy/laser/railgun/gauss/consume_next_projectile()
+	var/datum/component/heat/H = GetComponent(/datum/component/heat)
+	if((H.currentHeat > H.heatThresholdSpecial ||stored_matter < projectile_cost || !..()))
+		return null
+	stored_matter -= projectile_cost
+	return..()
+
+/obj/item/gun/energy/laser/railgun/gauss/consume_next_projectile()
+	if(stored_matter < projectile_cost) return null
+	if(!cell) return null
+	if(!ispath(projectile_type)) return null
+	if(!cell.checked_use(charge_cost)) return null
+	stored_matter -= projectile_cost
+	return new projectile_type(src)
```

Comment:
```suggestion
	return ..()

```
consume_next_projectile() is defined twice and as such will behave oddly (it will ignore heat)